src/hotspot/share/asm/register.hpp

```diff
@@ -93,67 +93,75 @@ template <class RegImpl> class ReverseRegSetIterator;
 // A set of registers
 template <class RegImpl>
 class AbstractRegSet {
-  uint32_t _bitset;
+  size_t _bitset;
```

Comment:
Why couple the number of possible registers to the memory size? Why not uint64_t?

Reply:
64-bit makes sense.

I think this had been broken for ppc where the number of vector registers can exceed 32 (https://github.com/openjdk/jdk/blob/91101f0d4fc8e06d0d74e06361db6ac87efeeb8e/src/hotspot/cpu/ppc/register_ppc.hpp#L378)

Unless I am mistaken, `assert_different_registers`, if applied to VSR32 and up, would never have fired.

Reply:
OK, I get it. I was sort-of thinking that on the 32-bit platforms we support we don't ever have more than 32 registers in a set, but maybe that's not true. I certainly don't want to slow down 32-bit platforms by burdening them with double-word operations for something that can never happen.

Reply:
Okay. Well, if we have a static assert, we will notice if we have more registers than fit the bitset.

Reply:
So I'm getting this bizarre failure on arm32. I'm guessing it's actually a compiler bug, but I suppose it might be some dusty corner of C++ to do with template arg substitution. Any thoughts?

```
In file included from /home/runner/work/jdk/jdk/src/hotspot/share/utilities/globalDefinitions.hpp:29,
                 from /home/runner/work/jdk/jdk/src/hotspot/share/nmt/memflags.hpp:28,
                 from /home/runner/work/jdk/jdk/src/hotspot/share/memory/allocation.hpp:29,
                 from ad_arm.hpp:30,
                 from ad_arm.cpp:28:
/home/runner/work/jdk/jdk/src/hotspot/share/asm/register.hpp: In instantiation of ‘class AbstractRegSet<RegisterImpl*>’:
/home/runner/work/jdk/jdk/src/hotspot/share/asm/register.hpp:272:30:   required from ‘constexpr bool different_registers(R, Rx ...) [with R = RegisterImpl*; Rx = {RegisterImpl*, RegisterImpl*}]’
/home/runner/work/jdk/jdk/src/hotspot/share/asm/register.hpp:278:27:   required from ‘void assert_different_registers(R, Rx ...) [with R = RegisterImpl*; Rx = {RegisterImpl*, RegisterImpl*}]’
/home/runner/work/jdk/jdk/src/hotspot/cpu/arm/arm.ad:8984:52:   required from here
/home/runner/work/jdk/jdk/src/hotspot/share/asm/register.hpp:96:26: error: ‘number_of_registers’ is not a member of ‘RegisterImpl*’
   96 |   STATIC_ASSERT(RegImpl::number_of_registers <= 64);
      |                          ^~~~~~~~~~~~~~~~~~~
/home/runner/work/jdk/jdk/src/hotspot/share/utilities/debug.hpp:287:44: note: in definition of macro ‘STATIC_ASSERT’
  287 | #define STATIC_ASSERT(Cond) static_assert((Cond), #Cond)
      |                                            ^~~~
```

Reply:
I just changed the static assert to a regular assert, in the hope that it might work better on arm32.

Reply:
Urgh. 

Seems on arm32 (and on zero, but I guess that does not matter) Register is a typedef to a pointer:

https://github.com/openjdk/jdk/blob/9686e804a2b058955ff88149c54a0a7896c0a2eb/src/hotspot/cpu/arm/register_arm.hpp#L136

and RegImpl is a pointer type, so it cannot have members.

On all other platforms, Register is a class.

I don't see a quick fix other than to make the assert dependent on !arm !zero. Or, leave it for a followup fix.

---

BTW, note that on Arm32, we can have 64 float registers:

https://github.com/openjdk/jdk/blob/9686e804a2b058955ff88149c54a0a7896c0a2eb/src/hotspot/cpu/arm/register_arm.hpp#L178

So, I would make the underlying type 64-bit for 32-bit builds too.

Reply:
BTW Would the runtime assert not prevent this from used as constexpr?

Reply:
It doesn't, and I just saw the problem: on Arm, `Register` is a pointer type, whereas on other ports it's a class type.

Reply:
Looks like our comments crossed over. I'm building Zero and it's OK so far.

Reply:
I guess Arm never got converted. That port needs some love.

Reply:
Or, deprecation.